sdk/identity/azure_identity/src/managed_identity_credential.rs

```diff
@@ -0,0 +1,498 @@
+                ));
+            }
+        };
+
+        let mut message = format!(
+            "ManagedIdentityCredential will use {} managed identity",
+            source.as_str()
+        );
+        if let Some(user_assigned_id) = options.user_assigned_id {
+            message.push_str(&format!(" with {}", user_assigned_id));
+        }
+        trace!(message);
+
+        Ok(Arc::new(Self { credential }))
+    }
+}
+
+#[cfg_attr(target_arch = "wasm32", async_trait::async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait::async_trait)]
+impl TokenCredential for ManagedIdentityCredential {
+    async fn get_token(&self, scopes: &[&str]) -> azure_core::Result<AccessToken> {
+        self.credential.get_token(scopes).await
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+enum ManagedIdentitySource {
+    AzureArc,
+    AzureML,
+    AppService,
+    CloudShell,
+    Imds,
+    ServiceFabric,
+}
+
+impl ManagedIdentitySource {
+    pub fn as_str(&self) -> &str {
+        match self {
+            ManagedIdentitySource::AzureArc => "Azure Arc",
+            ManagedIdentitySource::AzureML => "Azure ML",
+            ManagedIdentitySource::AppService => "App Service",
+            ManagedIdentitySource::CloudShell => "CloudShell",
+            ManagedIdentitySource::Imds => "IMDS",
+            ManagedIdentitySource::ServiceFabric => "Service Fabric",
+        }
+    }
+}
+
+const IDENTITY_ENDPOINT: &str = "IDENTITY_ENDPOINT";
+const IDENTITY_HEADER: &str = "IDENTITY_HEADER";
+const IDENTITY_SERVER_THUMBPRINT: &str = "IDENTITY_SERVER_THUMBPRINT";
+const IMDS_ENDPOINT: &str = "IMDS_ENDPOINT";
+const MSI_ENDPOINT: &str = "MSI_ENDPOINT";
+const MSI_SECRET: &str = "MSI_SECRET";
+
+fn get_source() -> ManagedIdentitySource {
+    use ManagedIdentitySource::*;
+    if env::var(IDENTITY_ENDPOINT).is_ok() {
+        if env::var(IDENTITY_HEADER).is_ok() {
+            if env::var(IDENTITY_SERVER_THUMBPRINT).is_ok() {
+                return ServiceFabric;
+            }
+            return AppService;
+        } else if env::var(IMDS_ENDPOINT).is_ok() {
+            return AzureArc;
+        }
+    } else if env::var(MSI_ENDPOINT).is_ok() {
+        if env::var(MSI_SECRET).is_ok() {
+            return AzureML;
+        }
+        return CloudShell;
+    }
+    Imds
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{LIVE_TEST_RESOURCE, LIVE_TEST_SCOPES};
+    use azure_core::http::headers::Headers;
+    use azure_core::http::{Method, Request, Response, StatusCode};
+    use azure_core::Bytes;
+    use azure_core_test::http::MockHttpClient;
+    use futures::FutureExt;
+    use serial_test::serial;
+    use std::collections::HashMap;
+    use std::time::{SystemTime, UNIX_EPOCH};
+
+    const EXPIRES_ON: &str = "EXPIRES_ON";
+
+    struct EnvVarGuard {
+        original_vars: HashMap<String, Option<String>>,
+    }
+
+    impl EnvVarGuard {
```

Comment:
`TokenCredentialOptions` already have this, which you should use to get variables at runtime and override at test time. You can pass in the options. This is also safer because we're not actually setting env vars in the process in case tests are ever parallelized (probably won't be, but one never knows).